flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterServiceLeadershipRunner.java

```diff
@@ -269,13 +269,17 @@ private void startJobMasterServiceProcessAsync(UUID leaderSessionId) {
     @GuardedBy("lock")
     private void verifyJobSchedulingStatusAndCreateJobMasterServiceProcess(UUID leaderSessionId)
             throws FlinkException {
-        final RunningJobsRegistry.JobSchedulingStatus jobSchedulingStatus =
-                getJobSchedulingStatus();
-
-        if (jobSchedulingStatus == RunningJobsRegistry.JobSchedulingStatus.DONE) {
-            jobAlreadyDone();
-        } else {
-            createNewJobMasterServiceProcess(leaderSessionId);
+        try {
+            if (jobResultStore.hasJobResultEntry(getJobID())) {
+                jobAlreadyDone();
+            } else {
+                createNewJobMasterServiceProcess(leaderSessionId);
+            }
+        } catch (IOException e) {
+            throw new FlinkException(
```

Comment:
There seems to be no reason to wrap these methods in `FlinkExceptions`. We'll wrap them in `RuntimeException` anyway in `#startJobMasterServiceProcessAsync()`. If we simply change the signature here and in `createNewJobMasterServiceProcess` to `IOException`, the code will simplify a lot.